starlette/tasks.py

```diff
@@ -0,0 +1,30 @@
+import uuid
+import asyncio
+import functools
+
+
+class BackgroundTask:
+    def __init__(self, func, *args, **kwargs):
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+
+    async def __call__(self):
+        if asyncio.iscoroutinefunction(self.func):
+            self._result = await asyncio.ensure_future(
+                self.func(*self.args, **self.kwargs)
+            )
+        else:
+            fn = functools.partial(self.func, *self.args, **self.kwargs)
+            loop = asyncio.get_event_loop()
+            self._result = await loop.run_in_executor(None, fn)
```

Comment:
Bit of a question here - we may, or may not want to do this.

For synchronous routes we don't currently run them in an executor (we just assume that you're only going to do some non-blocking stuff if you're not using `async`.) running in an executor here isn't consistent with *that* set of behavior, but it's possible that we want to switch things around anyways.

One option for now would be to *force* background tasks to only ever use `async` functions. If users want to run something in a thread pool then they can, explicitly, but we keep consistency better consistency with how we handle routes. (Not 100% either way on this tho)

Reply:
I think the option should exist either way. The docs could warn users that any sync tasks would be running in a thread pool. This allows more flexibility, especially if you need to use any third party code (most often synchronous) inside your background task.

I'd be interested to hear other opinions though. For now, would you just like me to force the `async` option only?

Reply:
I'm not sure, either we:

* Leave it as you have it now.
* Force it to `async` only, and at some point add threadpools for *both* background tasks, *and* for routes/websockets, making sure to support them everywhere, rather than in just one particular place.